File: python/ESElement.py
# -*- coding: utf-8 -*-
"""
Created on Sun Aug  1 13:24:02 2021

@author: philippe@loco-labs.io

This module contains the `ESElement` parent class of the `ES.ESObs.ESObs` and
`ES.ESObservation.Observation` classes. 
"""
import json
from ESconstante import ES

class ESElement:
    """
    
    *Attributes* :
    
    - **mAtt** : dict with properties of an instance
    - **typeES** : String, define the nature of an `ESElement`
    - **classES** : String, define the level of class
    - **pComposant** : list of `ESElement` included in the `ESElement`
    - **pContenant** : list of `ESElement` containing the `ESElement`
    
    The methods defined in this class are : 

    - `addComposant`
    - `element`
    - `getAttAll`
    
    
    - `isESAtt`(@staticmethod)
    - `isESObs`(@staticmethod)
    - `isUserAtt`(@staticmethod)
    
    
    """
    def __init__(self):
        self.mAtt = dict()
        self.typeES = "null"
        self.classES = "null"
        self.pComposant = list()
        self.pContenant = list()

    def __repr__(self):
        txt = object.__repr__(self) + '\n'
        txt += f"classES, typeES : {self.classES} {self.typeES} \n"
        for k, v in self.mAtt.items(): txt += f"{k} : {v} \n"
        if (len(self.pComposant) > 0): txt += f"nombre de composants {len(self.pComposant)} \n"
        if (len(self.pContenant) > 0): txt += f"nombre de contenants {len(self.pContenant)} \n"
        return txt
    
    def addComposant(self, pCompos):
        '''Link two `ESElement` 
        
        *Parameters*
        
        - **pCompos** : `ESElement` to include
        
        *Returns*
        
        - **None**
        '''
        self.pComposant.append(pCompos)
        pCompos.pContenant.append(self)

    def element(self, comp) :
        """Return `ESElement` included with a specific 'classES' or 'typeES'
        
        *Parameters*
        
        - **comp** : string to search
        
        *Returns*
        
        - **ESElement found**
        """
        for cp in self.pComposant:
            if (cp.typeES == comp or cp.classES == comp): return cp
            elif (cp.element(comp) != None): return cp.element(comp)
        return None

    def getAttAll(self, key):
        """Return mAtt value in a global hierachy
        
        *Parameters*
        
        - **key** : string to search
        
        *Returns*
        
        - **value found**
        """
        if self.isAtt(key): return self.mAtt[key];
        for compo in self.pComposant:
            if (compo.getAttAll(key) != ES.nullAtt): return compo.getAttAll(key)
        return ES.nullAtt

    @staticmethod
    def isESObs(esClass, jObj):
        """Identify if 'jObj' include 'esClass' keys
        
        *Parameters*
        
        - **jObj** : dict to be evaluated
        
        *Returns*
        
        - **Boolean**
        """        
        if type(jObj) == esClass : return True
        if type(jObj) != dict : return False
        for key, value in jObj.items():
            if key == esClass: return True
            for val, classES in ES.mValObs.items():
                if key == val and esClass == classES:  return True
            for val, classES in ES.mTypeAtt.items():
                if key == val and esClass == classES:  return True
        return False

    @staticmethod
    def isESAtt(esClass, key):
        """identify if 'key' is included in 'esClass'
        
        *Parameters*
        
        - **esClass** : string for an ESClass attribute
        - **key** : string to search
        
        *Returns*
        
        - **Boolean**
        """        
        for k,v in ES.mTypeAtt.items():
            if v == esClass and k == key: return True
        return False

    @staticmethod
    def isUserAtt(key):
        """identify if 'key' is included in any `ESElement` attribute
        
        *Parameters*
        
        - **key** : string to search
        
        *Returns*
        
        - **Boolean**
        """                
        for k,v in ES.mTypeAtt.items():
            if k == key: return False
        for k,v in ES.mValObs.items():
            if k == key: return False
        return True

    
    def _jsonAtt(self, **option):
        att = dict()
        for k, v in self.mAtt.items():
            if k in list(ES.mTypeAtt.keys()) : 
                if v not in ES.nullValues : att[k] = v
            else: att[k] = v
        return att

File: C++/ESElement.cpp
#include "stdafx.h"  // à supprimer arduino
#include <sstream>
#include "ESElement.h"
#include "ESObservation.h"

using namespace std;

const std::map<std::string, std::string> ESElement::mTypeAtt = {
{ "ResultSetTime",		"observation" },
{ "ResultSetQuality",	"ResultSet" },
{ "propertyType",	"PropertyValue" },
{ "EMFType",		"ObservingEMF" },
{ "ResultSetNature",	"ObservingEMF" },
{ "lowerValue",		"ObservingEMF" },
{ "upperValue",		"ObservingEMF" },
{ "period",			"ObservingEMF" },
{ "uncertainty",	"ObservingEMF" },
{ "unit",			"PropertyValue" },
{ "sampling",		"PropertyValue" },
{ "application",	"PropertyValue" }
};
const std::string ESElement::metaTypeESObject	= "ESObject";


ESElement::ESElement() { mAtt["type"] = "null"; typeES = "null"; classES = "null"; pContenant.clear(); pComposant.clear(); }
void ESElement::setAtt(std::string key, std::string value)	{ mAtt[key] = value; }
std::string ESElement::getAtt(std::string key) const		{ if (isAtt(key)) return  mAtt.at(key); return "null"; }
std::string ESElement::getTypeES() const					{ return typeES; }
std::string ESElement::getClassES() const					{ return classES; }
std::string ESElement::getMetaType() const					{ return metaType; }
std::map<std::string, std::string> ESElement::getmAtt() const { return mAtt; }
bool ESElement::isAtt(std::string key) const				{
	for (std::map<string, string>::const_iterator it = mAtt.begin(); it != mAtt.end(); ++it) if (it->first == key) return 1;
	return 0;
}
bool ESElement::isESAtt(std::string esClass, std::string key) {
	for (std::pair<string, string> couple : ESElement::mTypeAtt) if (couple.second == esClass and couple.first == key) return 1;
	return 0;
}
bool ESElement::isESObs(std::string esClass, JsonObject jObj) {
	bool esObs = 0;
	JsonObject objAtt = jObj[esClass];
	if (objAtt.isNull()) {
		for (JsonPair p : jObj) {
			if ((string)p.key().c_str() == esClass) esObs = 1;
			for (std::pair<string, string> couple : ESElement::mTypeAtt)
				if ((string)p.key().c_str() == couple.first and esClass == couple.second) esObs = 1;
		}
	} else esObs = 1;
	return esObs;
}
JsonObject ESElement::deserialize(std::string json) {
	const int capa = 1000;
	StaticJsonDocument<capa> doc;
	DeserializationError error = deserializeJson(doc, json);
	JsonObject obj = doc.as<JsonObject>();
	return obj;
}
std::string ESElement::getAttAll(std::string key) const {
	if (isAtt(key)) return mAtt.at(key);
	for (int i = 0; i < (int)pComposant.size(); i++) if (pComposant[i]->getAttAll(key) != "null") return pComposant[i]->getAttAll(key);
	return "null";
}
void ESElement::addComposant(ESElement* pCompos) {
	pComposant.push_back(pCompos);
	pCompos->pContenant.push_back(this);
}
void ESElement::println(std::string nam, std::string pr) { 
#ifdef TEST_ES
	cout << nam << " : " << pr << endl;					// standard
#else
	Serial.println((nam + " : " + pr).c_str());		// arduino
#endif
}
ESElement* ESElement::element(std::string comp) const {
	for (int i = 0; i < (int)pComposant.size(); i++) {
		if (pComposant[i]->getTypeES() == comp or pComposant[i]->getClassES() == comp or pComposant[i]->getMetaType() == comp or pComposant[i]->mAtt["type"] == comp)
			return pComposant[i];
		else if (pComposant[i]->element(comp) != nullptr) return pComposant[i]->element(comp);
	}
	return nullptr;
}
void ESElement::majMeta() { for (int i = 0; i < (int)pContenant.size(); i++) if (pContenant[i]->getTypeES() == Observation::ESclass) static_cast<Observation*>(pContenant[i])->majType(); }
void ESElement::print() const {
	std::stringstream ss;
	ESElement::println("classES, typeES : ", classES + " " + typeES);
	for (std::map<string, string>::const_iterator it = mAtt.begin(); it != mAtt.end(); ++it) ESElement::println(it->first, it->second);
	ss << pComposant.size();
	if (pComposant.size() > 0) ESElement::println("nombre de composants", ss.str());
	ss << pContenant.size();
	if (pContenant.size() > 0) ESElement::println("nombre de contenants", ss.str());
}
std::string ESElement::jsonAtt(bool complet) const {
	std::string json(""), deb, fin, firs, val;
	for (std::map<string, string>::const_iterator it = mAtt.begin(); it != mAtt.end(); ++it) {
		val = it->second;
		if ((complet or it->first[0] == '$') && (val != "null")) {
			if (it->first == "type" or it->first == "nval") firs = it->first + classES;
			else firs = it->first;
			deb = ""; fin = "";
			if (val[0] != '{' && val[0] != '[' && val[0] != '+' && val[0] != '-' && val[0] != '0' && val[0] != '1' && val[0] != '2' && val[0] != '3' && \
				val[0] != '4' && val[0] != '5' && val[0] != '6' && val[0] != '7' && val[0] != '8' && val[0] != '9' && val[0] != '"') {
				deb = "\""; fin = "\"";
			}
			json += "\"" + firs + "\":" + deb + val + fin + ",";
		}
	}
	return json;
}
ESObject::ESObject() : ESElement() { metaType = "ESObject"; name = "observation xx"; }

ESObs::ESObs() : ESElement() { metaType = "ESObs"; nValue = 0; }
ESObs::ESObs(Observation* pObs) : ESElement() { metaType = "ESObs"; nValue = 0; if (pObs != nullptr) pObs->addComposant(this); }
int ESObs::getNvalue() const { return nValue; }
void ESObs::setNvalue(int nval) { nValue = nval; stringstream nv; nv << nValue; mAtt["nval"] = nv.str(); }

